pallets/portfolio/src/lib.rs

```diff
@@ -329,4 +353,175 @@ impl<T: Trait> Module<T> {
+            // If a custodian is assigned, only they are allowed.
+            ensure!(
+                Self::portfolio_custodian(portfolio) == custodian,
+                Error::<T>::UnauthorizedCustodian
+            );
+        } else {
+            // Else, only the portfolio owner is allowed
+            // TODO: support portfolio permissions
+            ensure!(
+                portfolio.did == custodian,
+                Error::<T>::UnauthorizedCustodian
+            );
+        }
+
+        Ok(())
+    }
+
+    pub fn check_portfolio_transfer_validity(
+        from_portfolio: PortfolioId,
+        to_portfolio: PortfolioId,
+        amount: <T as CommonTrait>::Balance,
+        ticker: &Ticker,
+    ) -> DispatchResult {
+        // 1. Ensure from and to portfolio are different
+        ensure!(
+            from_portfolio != to_portfolio,
+            Error::<T>::SelfTransferNotAllowed
+        );
+
+        // 2. Ensure sender has enough free balance
+        let from_balance = Self::portfolio_asset_balances(&from_portfolio, ticker);
+        ensure!(
+            from_balance
+                .saturating_sub(Self::locked_assets(&from_portfolio, ticker))
+                .checked_sub(&amount)
+                .is_some(),
+            Error::<T>::InsufficientPortfolioBalance
+        );
+
+        Ok(())
+    }
+
+    pub fn lock_tokens(
+        portfolio: PortfolioId,
+        amount: <T as CommonTrait>::Balance,
+        ticker: &Ticker,
+    ) -> DispatchResult {
+        // 1. Ensure portfolio has enough free balance
+        let balance = Self::portfolio_asset_balances(&portfolio, ticker);
+        ensure!(
+            balance
+                .saturating_sub(Self::locked_assets(&portfolio, ticker))
+                .checked_sub(&amount)
+                .is_some(),
+            Error::<T>::InsufficientPortfolioBalance
+        );
+
+        // 2. Lock tokens.
+        // Locks are stacked so if there were X tokens already locked, there will now be X + N tokens locked
+        <PortfolioLockedAssets<T>>::mutate(&portfolio, ticker, |locked| {
+            *locked = locked.saturating_add(amount)
+        });
+        Ok(())
+    }
+
+    pub fn unlock_tokens(
+        portfolio: PortfolioId,
+        amount: <T as CommonTrait>::Balance,
+        ticker: &Ticker,
+    ) -> DispatchResult {
+        // 1. Ensure portfolio has enough locked tokens
+        let locked = Self::locked_assets(&portfolio, ticker);
+        ensure!(locked >= amount, Error::<T>::InsufficientTokensLocked);
+
+        // 2. Unlock tokens. Can not underflow due to above ensure.
+        <PortfolioLockedAssets<T>>::insert(&portfolio, ticker, locked - amount);
+        Ok(())
+    }
+}
+
+impl<T: Trait> PortfolioSubTrait<T::Balance> for Module<T> {
+    fn accept_portfolio_custody(new_custodian: IdentityId, auth_id: u64) -> DispatchResult {
+        Self::accept_portfolio_custody(new_custodian, auth_id)
```

Comment:
Why not just have the definitions here?

Reply:
That makes sense. It breaks the pattern but for the better, I suppose.